ChoiceView - Add support for 'optional' or 'no-selection'

jlhughes opened the pull request:
If the ChoiceView is 'optional' then the 'placeholder' is presented as the first list entry and if an explicit choice is not set, the 'placeholder' is selected. The ChoiceView returns null for the placeholder selection.

jlhughes committed:
Add support for 'optional' or 'no-selection'

If the ChoiceView is 'optional' then the 'placeholder' is presented as the first list entry and if an explicit choice is not set, the 'placeholder' is selected. The ChoiceView returns null for the placeholder selection.

kgrgreer commented on src/foam/u2/tag/Select.js:
```diff
@@ -45,8 +47,15 @@ foam.CLASS({
+                selected: self.data == -1 ? true : undefined,
+                value: -1
+            }).add(self.placeholder));
+          } else {
```
Your indentation makes it look like the else is for the first if statement, but it's actually for the inner one.

kgrgreer commented:
Is there any case where you would would not want it to be 'optional'? Is this needed, or can we just change the default behaviour?

jlhughes committed:
Select - Correct indentation

jlhughes commented:
I'm unable to 'comment' directly on the pull-request at the moment.

re: optional - default behaviour
I have a roughly 70/30 split between optional/non-optional for the 20 or so
ChoiceViews.  For me I think of 'enabling' optional rather than 'disabling'
it.   Also, for your examples and use-cases you've had no need for
optional.

On Tue, Feb 7, 2017 at 10:20 AM, Kevin Glen Roy Greer <
notifications@github.com> wrote:

> Is there any case where you would would not want it to be 'optional'? Is
> this needed, or can we just change the default behaviour?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/foam-framework/foam2-experimental/pull/234#issuecomment-278031519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AApv82a42o6_MbM2-qNC6t8Hrp_T3hQWks5raIvHgaJpZM4L3kx1>
> .
>

jlhughes committed:
Merge remote-tracking branch 'upstream/master' into ChoiceView-Optional

jlhughes committed:
Refactor optional or autoselect using just the placeholder property.

By defaulting the placeholder property to undefined it can act as both
a flag to enable the use of the placeholder in the selection list and
the text value of the placeholder selection list entry.

jlhughes committed:
Remove unused 'optional' property

jlhughes committed:
Select - remove trailing comma

jlhughes committed:
Select - adjust indenting
